### ISSUE TITLE
Nova Flexible Content

### ISSUE THREAD

**Issue by dividy** [dcasia/conditional-container#17]
Hello,

Is it compatible with Nova Flexible Content ?

I tried it and had a JS error..

Thanks !

**Comment by milewski** [comment]
Currently it doesn't work with flexible content

**Comment by dividy** [comment]
Yep, that's why i tried Nova Dependency Container but I encountered 2
issues with that package too :

- The conditional field doesn't show on the detail view

- Some values are overwritten by the last item of the layout


I've seen you used that package too, do these issues ring a bell to you ?

Thanks.

Le sam. 14 déc. 2019 à 02:34, Rafael Milewski <notifications@github.com> a
écrit :

**Comment by milewski** [comment]
Yeah I used that package before too but had too many issues for my use case so I built this one instead, I have seen how the `Nova Dependency Container` implemented support for the flexible content on this commit: https://github.com/epartment/nova-dependency-container/commit/d4081499c45b86111d5d95ef3c4457db4dd4bbdf perhaps it may be very simple to port it over... I will give it a shoot, and about the 2 issues, I don't see how `conditional-container` would fail in there

**Comment by milewski** [comment]
@dividy I got it almost working and I had pushed my proof of concept to a branch called flexible field, it works on create, on update and on details view as expected, however validation is still broken :(

**Comment by dividy** [comment]
Wow.. This sounds like a great step forward to something functional ;-)
I'll try to have a look at this too!

**Comment by dividy** [comment]
Just did a **composer require digital-creative/conditional-container:dev-flexible-content**

Tried the package, and I'm still having an issue inside Flexible Content...

Here's the code, can you try it on your side and tell me if it works ?

``
Flexible::make('Produits', 'products')
            ->addLayout('Product', 'product', [
                Select::make('Type de contrat', 'type')->options([
                    'L' => 'Location',
                    'VL' => 'Vente & Location',
                    'R' => 'Relevé',
                    'V' => 'Vente',
                    'VU' => 'Vente unique',
                ])->displayUsingLabels()->rules('required'),
    
                ConditionalContainer::make([ Text::make('Prix en Location (htva)', 'price_rent') ])
                ->if('type = L'),
              
                ConditionalContainer::make([ Text::make('Prix de Vente (htva)', 'price_sale') ])
                ->if('type = V')
                ->if('type = VU'), 
    
                ConditionalContainer::make([ 
                    Text::make('Prix en Location (htva)', 'price_rent'),
                    Text::make('Prix de Vente (htva)', 'price_sale')
                ])->if('type = VL'),
    
                ConditionalContainer::make([ 
                    Text::make('Prix du Relevé (htva)', 'price_reading'),
                ])->if('type = R'),        
            ]),

**Comment by dividy** [comment]
On my side, it never works, and displays all the fields when I select 'R - Relevé'.. Strange ;)

**Comment by bernhardh** [comment]
@milewski  I just used the branch as well and tried the code of @dividy but i get 

```
TypeError: Cannot read property 'join' of undefined
    at _loop (conditional-container:175)
```

And I even dont see the fields.

I also tried my own, simpler example with the same result:

```php
Flexible::make("Data")
    ->addLayout("Simple Layout", "html", [
        Select::make("Type", "type")->options([
            "html" => "HTML",
            "plain" => "Plaintext"
        ]),
        ConditionalContainer::make([
            Trix::make('Content', 'html_content')->gridClass("w-full")
        ])->if('type = html'),
        ConditionalContainer::make([
            Textarea::make('Content', 'html_content')->gridClass("w-full")
        ])->if('type = plain"')
    ])
```

**Comment by milewski** [comment]
@bernhardh you probably forgot to include the `HasConditionalContainer` trait your example works for me:

![image](https://user-images.githubusercontent.com/2874967/72213344-50df5900-3528-11ea-88ad-5bcc195a952f.png)

**Comment by milewski** [comment]
The `join` issue is because the `HasConditionalContainer` wasn't included ... but yeah indeed there are several issues, it did work for the example I had when I was developing it but using your samples has some bugs...

**Comment by milewski** [comment]
@bernhardh I just fixed the issues with your example you can try again it should be working perfectly now (but without validation of course)

# Form View
![image](https://user-images.githubusercontent.com/2874967/72213692-17115100-352e-11ea-9cf1-5d0453f019a0.png)

# Detail View
![image](https://user-images.githubusercontent.com/2874967/72213698-285a5d80-352e-11ea-9d2d-bd9844849162.png)

**Comment by milewski** [comment]
@dividy your example also works now for me... I think the issue were you forgot the trait

**Comment by bernhardh** [comment]
Oh. So stupid. Thank you!

Now I get 

```
Method Whitecube\NovaFlexibleContent\Flexible::resolveConditionalContainer does not exist.
```

The 

```php
if ($flexibleContent->isNotEmpty()) {
    $this->registerFlexibleMacros($request, $flexibleContent);
}
```

is false, so this macro is never added?

**Comment by milewski** [comment]
whats the endpoint nova is trying to hit that throws that error `Method Whitecube\NovaFlexibleContent\Flexible::resolveConditionalContainer does not exist.` ?

**Comment by bernhardh** [comment]
It occurs on edit and detail page. And the endpoint throwing it is `/nova-api/MODELNAME/ID`

**Comment by bernhardh** [comment]
Btw. I am on `dev-flexible-content 7e5e620` atm.

**Comment by milewski** [comment]
Oh I see when you use the nova-tabs the "flexible" content is not found, so the macros are not added to it, that's why it crashes, I think it might be possible to detect, I think it is possible to add a check and see if its an instance of panel and recursively try to find flexible contents within its children

**Comment by bernhardh** [comment]
Ok, at least I can fix the issue with the tabs package (and all other packages based on `Panel` items).

In HasConditionalContainer.php:460 just add

``php
                if($field instanceof Panel && $field->data) {
                    return $this->findAllFlexibleContentFields($field->data);
                }
```

**Comment by milewski** [comment]
Ah ok let me add

**Comment by milewski** [comment]
Pushed

**Comment by bernhardh** [comment]
Ok, now it only worked on empty data. After save i get 

```
Cannot bind an instance to a static closure
```

To fix this change `HasConditionalContainer:178`

```php
$field::macro('generateFieldName', static function (array $fields) {
```

to

```php
$field::macro('generateFieldName', function (array $fields = []) {
```

Not sure why it is decleared static in the first place? Or does this have any sideeffects?

**Comment by milewski** [comment]
That was a mistake, I didn't know macros wouldnt run in static context, I have removed it and pushed again

**Comment by bernhardh** [comment]
Thank you very much, for your help, this package (and `nova-mega-filter` and `collapsible-resource-manager`). Do you have a "buy me a cup of coffee" link or something else?

**Comment by dividy** [comment]
Good job guys, I watched all the action from far away, getting spammed by notifications.

I have to deliver the application to my customer, so I finally chose another working option, I used different presets instead of conditional fields. However, great work!

**Comment by patrickleemsantos** [comment]
@dividy Can I know what you use in order to fix this issue?

**Comment by anditsung** [comment]
> @dividy I got it almost working and I had pushed my proof of concept to a branch called flexible field, it works on create, on update and on details view as expected, however validation is still broken :(

```
if ($controller instanceof CreationFieldController ||
            $controller instanceof UpdateFieldController) {
```

this is the part where you check the fields for creation?
